src/Aws/Kinesis/Client/Consumer.hs

```diff
@@ -241,93 +220,92 @@ managedKinesisConsumer kit =
+        threadDelay $
+          case recordsCount of
+            0 → 5000000
+            _ → 70000
+
+  withAsync (liftIO reshardingLoop) $ \reshardingHandle → do
+    link reshardingHandle
+    withAsync (liftIO producerLoop) $ \producerHandle → do
+      link producerHandle
+      res ← inner $ KinesisConsumer messageQueue state
+      return res
 
 -- | This requests new information from Kinesis and reconciles that with an
 -- existing carousel of shard states.
 --
 updateStreamState
-  ∷ MonadConsumerInternal m
-  ⇒ StreamState
-  → m StreamState
-updateStreamState state = do
-  streamName ← view ckStreamName
-  defaultIteratorType ← view ckIteratorType
-  savedState ← view ckSavedStreamState
-
-  mapError KinesisError ∘ mapEnvironment ckKinesisKit $ do
-    let existingShardIds = state ^. CR.clList <&> view ssShardId
-        shardSource = flip mapOutputMaybe (streamOpenShardSource streamName) $ \sh@Kin.Shard{..} →
-          if shardShardId `elem` existingShardIds
-            then Nothing
-            else Just sh
-
-    newShards ← shardSource $$ CL.consume
-    shardStates ← forM newShards $ \Kin.Shard{..} → do
-      let startingSequenceNumber =
-            savedState ^? _Just ∘ _SavedStreamState ∘ ix shardShardId
-          iteratorType =
-            maybe
-              defaultIteratorType
-              (const Kin.AfterSequenceNumber)
-              startingSequenceNumber
-
-      Kin.GetShardIteratorResponse it ← runKinesis Kin.GetShardIterator
-        { Kin.getShardIteratorShardId = shardShardId
-        , Kin.getShardIteratorShardIteratorType = iteratorType
-        , Kin.getShardIteratorStartingSequenceNumber = startingSequenceNumber
-        , Kin.getShardIteratorStreamName = streamName
+  ∷ ConsumerKit
+  → StreamState
+  → IO StreamState
+updateStreamState ConsumerKit{..} state = do
+  let
+    existingShardIds = state ^. CR.clList <&> view ssShardId
+    shardSource =
+      flip mapOutputMaybe (streamOpenShardSource _ckKinesisKit _ckStreamName) $ \sh@Kin.Shard{..} →
+        if shardShardId `elem` existingShardIds
+          then Nothing
+          else Just sh
+
+  newShards ← shardSource $$ CL.consume
+  shardStates ← forM newShards $ \Kin.Shard{..} → do
+    let
+      startingSequenceNumber =
+        _ckSavedStreamState ^? _Just ∘ _SavedStreamState ∘ ix shardShardId
+      iteratorType =
+        maybe
+          _ckIteratorType
+          (const Kin.AfterSequenceNumber)
+          startingSequenceNumber
+
+    Kin.GetShardIteratorResponse it ← runKinesis _ckKinesisKit Kin.GetShardIterator
+      { Kin.getShardIteratorShardId = shardShardId
+      , Kin.getShardIteratorShardIteratorType = iteratorType
+      , Kin.getShardIteratorStartingSequenceNumber = startingSequenceNumber
+      , Kin.getShardIteratorStreamName = _ckStreamName
+      }
+    liftIO $ do
+      iteratorVar ← newTVarIO $ Just it
+      sequenceNumberVar ← newTVarIO Nothing
+      return ShardState
+        { _ssIterator = iteratorVar
```

Comment:
This pattern looks interesting. I try to understand why a new `TVar` is used here instead of just updating the existing var. 

* Is the content of the reference updated somewhere in the code? Otherwise maybe a immutable value could be used instead?
* Are there users of the old reference after this field got updated with a new reference?